uqcsbot/scripts/leet.py

```diff
@@ -0,0 +1,167 @@
+    collect_questions(questions)
+
+    # No difficulty specified, randomly generate one
+    if (len(difficulty) == 0):
+        difficulty = random.choice(LC_DIFFICULTY_MAP) 
+    
+    
+    selected_question = select_question(difficulty, questions) # Get a random question
+
+    # If we didn't find any questions for this difficulty, try again, probably timeout on all 3
+    if (len(selected_question) == 0):
+        bot.post_message(command.channel_id, 
+            "Hmm, the internet pipes are blocked. Try that one again.")
+        return
+
+    # Leetcode difficulty colors
+    if (difficulty == "easy"):
+        color = EASY_COLOR_CODE # Could make these constants, but they're just ripped from LC
+    elif (difficulty == "medium"):
+        color = MED_COLOR_CODE
+    else:
+        color = HARD_COLOR_CODE
+
+    
+
+    if (wasRandom):
+        title_text = f"Random {difficulty} question generated!"
+    else:
+        # Style this a bit nicer
+        difficulty = difficulty.title()
+        title_text = f"{difficulty} question generated!"
+    
+    difficulty = difficulty.title() # If we haven't already (i.e. random question)
+
+    msg_text = f"Here's a new question for you! <{selected_question[1]}|{selected_question[0]}>"
+
+    bot.post_message(command.channel_id, text=title_text,
+                    attachments=[Attachment(SectionBlock(msg_text), color=color)._resolve()])
+
+
+
+def select_question(difficulty, questions):
+    if (len(questions[difficulty]) == 0):
+        return ()
+    return random.choice(questions[difficulty])
+
+
+def collect_questions(questions):
+    """
+    Helper method to send GET requests to various Leetcode and HackerRank APIs. 
+    Populates provided dict (urls) with any successfully retrieved data, 
+    in the form of (Question_Title, Question_Link) tuple pairs.
+    """
+
+    # Get all the questions off the internet: hr data struct, hr algo, all leetcode
+    try: 
+        hr_ds = requests.get(HR_DS_API_LINK, headers={'User-Agent': 'Mozilla/5.0'}, 
+            timeout = 3)
+    except (requests.exceptions.ReadTimeout, requests.exceptions.ConnectTimeout) as error:
+        print("HackerRank Data Structure API timed out!" + "\n" + str(error))
+        hr_ds = None # If any of these fail (timeout etc) make them None for catching later
+
+    try:
+        hr_alg = requests.get(HR_ALG_API_LINK, headers={'User-Agent': 'Mozilla/5.0'}, timeout = 3)
+    except (requests.exceptions.ReadTimeout, requests.exceptions.ConnectTimeout) as error:
+        print("Hackerrank Algorithms API timed out!" + "\n" + str(error))
+        hr_alg = None
+
+    try:
+        lc_all = requests.get(LC_API_LINK, timeout=3)
+    except (requests.exceptions.ReadTimeout, requests.exceptions.ConnectTimeout) as error:
+        print("Leetcode API timed out!" + "\n" + str(error))
+        lc_all = None
+    
+    
+    # Parse the JSON blobs grabbed; if bad response, make dummy values to not include
```

Comment:
Same as above, blocks are all the same so you can just iterate. Obvs less good given the extra work for leetcode.

Reply:
Implemented